### PR TITLE
Allow return types, variadic and byref

### PR DESCRIPTION
**Pull request by hboomsma**
some non functional tests should still be added.

### CODE REVIEW

File: src/ReflectionMethod.php [hunk 1]
```diff
@@ -41,12 +49,82 @@ public function isStatic()
...
     }
 
+    public function getReturnType()
+    {
+        if (!method_exists($this->method, 'getReturnType')) {
+            return null;
+        }
+
+        if ($type = $this->method->getReturnType()) {
+            // Self is not valid then used in different places.
```
Comment (nicoschoenmaker) [inline src/ReflectionMethod.php]: Weird sentence, should it be 'when' instead of 'then'?

File: src/ReflectionMethod.php [hunk 2]
```diff
@@ -41,12 +49,82 @@ public function isStatic()
...
+    {
+        $context_factory = new ContextFactory();
+        $context         = $context_factory->createFromReflector($this->method);
+        $fqn_resolver    = new FqsenResolver();
+
+        try {
+            $resolved_type = $fqn_resolver->resolve($type, $context);
+        } catch (\InvalidArgumentException $e) {
+            print_r($e);
+            return '>>>WHOOPS: ' . $type;
```
Comment (nicoschoenmaker) [inline src/ReflectionMethod.php]: Why catch this exception? Seems to me like a nice way to silently ignore the error, and make sure we forget the stack trace :wink:

File: src/ReflectionParameter.php [hunk 3]
```diff
@@ -0,0 +1,81 @@
+<?php
+namespace Hostnet\Component\EntityPlugin;
+
+class ReflectionParameter
```
Comment (nicoschoenmaker) [inline src/ReflectionParameter.php]: Classdoc, why do we wrap the ReflectionParameter?

File: test/ReflectionGeneratorTest.php [hunk 4]
```diff
@@ -101,18 +105,97 @@ public function testGenerateWithMissingParent()
...
         $package_class = new PackageClass('\stdClass', __DIR__);
         $package_io    = self::createMock(Filesystem::class);
 
         $package_io->expects($this->once())
-               ->method('dumpFile')
-               ->with(
-                   dirname(__DIR__) . '/Generated/stdClassInterface.php',
-                   $this->matchesRegularExpression('/interface stdClassInterface/')
-               );
+                   ->method('dumpFile')
```
Comment (nicoschoenmaker) [inline test/ReflectionGeneratorTest.php]: Indent was already too far to the right, now ever further?

File: src/ReflectionMethod.php (hunk 1 above)
Comment (hboomsma) [inline src/ReflectionMethod.php]: :heavy_check_mark:

File: src/Resources/templates/parameters.php.twig [hunk 5]
```diff
@@ -1,9 +1,18 @@
...
+{% for parameter in parameters -%}
+    {% if parameter.hasType -%}
+        {% if (not parameter.isOptional() and parameter.allowsNull())
+           or (
+            parameter.allowsNull()
+            and parameter.isDefaultValueAvailable()
+            and parameter.defaultValue is not null) -%}
+            ?
+        {%- endif -%}
+        {{ parameter.type.name }} {# -#}
```
Comment (nicoschoenmaker) [inline src/Resources/templates/parameters.php.twig]: Commented code?

File: src/Resources/templates/parameters.php.twig [hunk 6]
```diff
@@ -1,9 +1,18 @@
...
-{{ type_hinter.typeHint(parameter) -}}
-{# TODO add typehint for class. Cant call parameter.class because that will trigger the autoloader #}
-${{  parameter.name }}
-{%- if parameter.isOptional() %}
-{%- if parameter.allowsNull() %} = null{% else %} = {{ parameter.defaultValue }}{% endif %}
-{%- endif %}
-{% if not loop.last %}, {% endif %}
+{% for parameter in parameters -%}
+    {% if parameter.hasType -%}
+        {% if (not parameter.isOptional() and parameter.allowsNull())
```
Comment (nicoschoenmaker) [inline src/Resources/templates/parameters.php.twig]: Really complicated if statement for a template. Can we add a function on parameter, so it's easier to see the intent behind this code?

File: test/Functional/src/Entity/VariadicTypedParameters.php [hunk 7]
```diff
@@ -0,0 +1,9 @@
+<?php
+namespace Hostnet\FunctionalFixtures\Entity;
+
+if (!defined('HHVM_VERSION')) {
```
Comment (nicoschoenmaker) [inline test/Functional/src/Entity/VariadicTypedParameters.php]: We dropped HHVM right?

File: src/Resources/templates/parameters.php.twig (hunk 5 above)
Comment (hboomsma) [inline src/Resources/templates/parameters.php.twig]: no not actual comment... it has a side effect

Comment (nicoschoenmaker) [inline src/Resources/templates/parameters.php.twig]: `{{ ' ' -}}`

Comment (nicoschoenmaker) [inline src/Resources/templates/parameters.php.twig]: `{{ parameter.type.name ~ ' ' -}}`

File: src/Resources/templates/parameters.php.twig (hunk 6 above)
Comment (nicoschoenmaker) [inline src/Resources/templates/parameters.php.twig]:
```twig
{% if parameter.allowsNull() %}
  {% set is_obligatory = not parameter.isOptional() %}
  {% set has_actual_default_value = parameter.isDefaultValueAvailable() and parameter.defaultValue is not null %}
  {% if is_obligatory or has_actual_default_value %}
    ?
  {% endif %}
{% endif %}
```

File: src/Resources/templates/parameters.php.twig (hunk 5 above)
Comment (hboomsma) [inline src/Resources/templates/parameters.php.twig]: fixed

File: test/Functional/src/Entity/VariadicTypedParameters.php (hunk 7 above)
Comment (hboomsma) [inline test/Functional/src/Entity/VariadicTypedParameters.php]: fixed

File: src/Resources/templates/parameters.php.twig (hunk 6 above)
Comment (hboomsma) [inline src/Resources/templates/parameters.php.twig]: fixed

**Comment by hboomsma** [comment]
will fix broken 56 test after lunch...

**Pull request merged**